Top level error handling

nnethercote opened the pull request:
The interactions between the following things are surprisingly complicated:
- `emit_stashed_diagnostics`,
- `flush_delayed`,
- normal return vs `abort_if_errors`/`FatalError.raise()` unwinding in the call to the closure in `interface::run_compiler`.

This PR disentangles it all.

r? @oli-obk

nnethercote committed:
Remove an out-of-date comment.

nnethercote committed:
Remove dead `expect_error_or_delayed_bug` method.

nnethercote committed:
Remove `EarlyDiagCtxt::abort_if_errors`.

Its one use isn't necessary, because it's not possible for errors to
have been emitted at that point.

nnethercote commented:
(Very rough draft code, please ignore for now.)

bors commented:
:umbrella: The latest upstream changes (presumably #121240) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote committed:
Adjust the `has_errors*` methods.

Currently `has_errors` excludes lint errors. This commit changes it to
include lint errors.

The motivation for this is that for most places it doesn't matter
whether lint errors are included or not. But there are multiple places
where they must be includes, and only one place where they must not be
included. So it makes sense for `has_errors` to do the thing that fits
the most situations, and the new `has_errors_excluding_lint_errors`
method in the one exceptional place.

The same change is made for `err_count`. Annoyingly, this requires the
introduction of `err_count_excluding_lint_errs` for one place, to
preserve existing error printing behaviour. But I still think the change
is worthwhile overall.

nnethercote committed:
Overhaul the handling of errors at the top-level.

Currently `emit_stashed_diagnostic` is called from four(!) different
places: `print_error_count`, `DiagCtxtInner::drop`, `abort_if_errors`,
and `compile_status`.

And `flush_delayed` is called from two different places:
`DiagCtxtInner::drop` and `Queries`.

This is pretty gross! Each one should really be called from a single
place, but there's a bunch of entanglements. This commit cleans up this
mess.

Specifically, it:
- Removes all the existing calls to `emit_stashed_diagnostic`, and adds
  a single new call in `finish_diagnostics`.
- Removes the early `flush_delayed` call in `codegen_and_build_linker`,
  replacing it with a simple early return if delayed bugs are present.
- Changes `DiagCtxtInner::drop` and `DiagCtxtInner::flush_delayed` so
  they both assert that the stashed diagnostics are empty (i.e.
  processed beforehand).
- Changes `interface::run_compiler` so that any errors emitted during
  `finish_diagnostics` (i.e. late-emitted stashed diagnostics) are
  counted and cannot be overlooked. This requires adding
  `ErrorGuaranteed` return values to several functions.
- Removes the `stashed_err_count` call in `analysis`. This is possible
  now that we don't have to worry about calling `flush_delayed` early
  from `codegen_and_build_linker` when stashed diagnostics are pending.
- Changes the `span_bug` case in `handle_tuple_field_pattern_match` to a
  `delayed_span_bug`, because it now can be reached due to the removal
  of the `stashed_err_count` call in `analysis`.
- Slightly changes the expected output of three tests. If no errors are
  emitted but there are delayed bugs, the error count is no longer
  printed. This is because delayed bugs are now always printed after the
  error count is printed (or not printed, if the error count is zero).

There is a lot going on in this commit. It's hard to break into smaller
pieces because the existing code is very tangled. It took me a long time
and a lot of effort to understand how the different pieces interact, and
I think the new code is a lot simpler and easier to understand.

nnethercote committed:
Inline and remove `Session::compile_status`.

Because it's now simple enough that it doesn't provide much benefit.

nnethercote committed:
Replace unnecessary `abort_if_errors`.

Replace `abort_if_errors` calls that are certain to abort -- because
we emit an error immediately beforehand -- with `FatalErro.raise()`.

nnethercote committed:
Refactor `run_global_ctxt`.

It currently is infallible and uses `abort_if_errors` and
`FatalError.raise()` to signal errors. It's easy to instead return a
`Result<_, ErrorGuaranteed>`, which is the more usual way of doing
things.

nnethercote committed:
Inline and remove `abort_on_err`.

It's clumsy and doesn't improve readability.

oli-obk commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -801,8 +804,6 @@ impl DiagCtxt {
     pub fn print_error_count(&self, registry: &Registry) {
         let mut inner = self.inner.borrow_mut();
 
-        inner.emit_stashed_diagnostics();
```
Either add an assertion here, too, or leave a comment as to why there may be stashed, unemitted diagnostics.

oli-obk commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -724,7 +724,7 @@ fn link_dwarf_object<'a>(
         Ok(()) => {}
         Err(e) => {
             sess.dcx().emit_err(errors::ThorinErrorWrapper(e));
-            sess.dcx().abort_if_errors();
+            FatalError.raise();
```
could also just change the `emit_err` to `emit_fatal`, right?

oli-obk commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -938,7 +938,7 @@ pub fn start_codegen<'tcx>(
         if let Err(error) = rustc_mir_transform::dump_mir::emit_mir(tcx) {
             let dcx = tcx.dcx();
             dcx.emit_err(errors::CantEmitMIR { error });
-            dcx.abort_if_errors();
+            FatalError.raise();
```
similar here: `emit_fatal`

oli-obk commented on compiler/rustc_codegen_ssa/src/base.rs:
```diff
@@ -451,8 +452,7 @@ pub fn maybe_create_entry_wrapper<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
             let dcx = cx.tcx().dcx();
             dcx.emit_err(errors::MultipleMainFunctions { span });
-            dcx.abort_if_errors();
-            bug!();
+            FatalError.raise();
```
and here

oli-obk reviewed (changes requested):
r=me with some smaller things fixed

klensy commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -1289,8 +1299,9 @@ impl DiagCtxtInner {
                     continue;
                 }
             }
-            self.emit_diagnostic(diag);
```
This eagerly calls `emit_diagnostic`, looks suspicious.

nnethercote replied:
Sorry, I don't understand this. How is it suspicious? What is the alternative to calling it "eagerly"?

klensy replied:
Sorry, linked wrong line.

`guar = guar.or(self.emit_diagnostic(diag));`

If i see `or` with `Option`, expecting that `or`'s argument will be 1. cheap (for perf reasons, not applied here) 2. without side effects, otherwise using `or` is misleading: you eval arg, but throwing out its result.

nnethercote replied:
I see. How would you write it instead?

nnethercote replied:
Because the idea here is that every `ErrorGuaranteed` is equivalent. So it doesn't matter which one we use. And it also relies on `or` not short-circuiting.

klensy replied:
Other problem is that in future someone will see that function call in .or(..) and will refactor into .or_else(..) which will break that `not short-circuiting` expectation here.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit a094903e9cdd06e92cb3f15f8d2fc9dbfc491231 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk a094903e9cdd06e92cb3f15f8d2fc9dbfc491231 -->
<!-- homu: {"type":"Approved","sha":"a094903e9cdd06e92cb3f15f8d2fc9dbfc491231","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #120576) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased.

@bors r=oli-obk

bors commented:
:pushpin: Commit 21827b9f6f884a98dd33ca3249f3f64453cba3d4 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 21827b9f6f884a98dd33ca3249f3f64453cba3d4 -->
<!-- homu: {"type":"Approved","sha":"21827b9f6f884a98dd33ca3249f3f64453cba3d4","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #121400) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased again.

@bors r=oli-obk

bors commented:
:pushpin: Commit f16c226af3ad4d21021433745214c71e95f10236 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk f16c226af3ad4d21021433745214c71e95f10236 -->
<!-- homu: {"type":"Approved","sha":"f16c226af3ad4d21021433745214c71e95f10236","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
